compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -613,6 +618,57 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             }),
         }
     }
+
+    /// Detect the inference regression on crate `time` <= 0.3.35 and emit a more targeted error.
+    /// <https://github.com/rust-lang/rust/issues/127343>
+    // FIXME: we should figure out a more generic version of doing this, ideally in cargo itself.
```

Comment:
Please leave a FIXME to remove this in a few versions when the time-0.3.35 and below's downloads are lower.

I very much do not want this hack to remain in the compiler for very long. IMO all crate-related hacks in the compiler (even the very old ones, like the rental and bevy hacks) should have some sort of "phasing out" condition.

Reply:
We have discussed having these kind of checks being done by cargo using some kind of database/file mechanism a-la RustSect for a while. I think that's got to be the future of these (which is why I mention it in the FIXME). Once we do that we no longer have to keep these kind of hacks in the compiler itself and it gets divorced from how many downloads the crate has (trying to compile a project that uses `time = "0.3.34"` 5 years from now, I'd appreciate having *some* hint of what's broken).

Reply:
This is not the first nor last time that we have broken a dependency due to an impl, and honestly I'd rather we *not* special case `time`. Nobody is committing to the follow-up work of cleaning up this hack, and IMO this simply leaves the compiler with one more special case that will continue to bitrot as time goes on.

Reply:
The fact that we, as T-compiler, need to maintain an ever-growing set of diagnostics hacks for things like this makes me pretty uncomfortable, expecially because this:

> We have discussed having these kind of checks being done by cargo using some kind of database/file mechanism a-la RustSect for a while.

...is just an *idea*, so we cannot be relying it on some future possibility as the removal condition of this hack.

Reply:
Resolving as this is now implemented under a hard time-gate which will block compilation if it's not removed before bumping to 1.89.